vertical-pod-autoscaler/enhancements/3919-customized-recommender-vpa/README.md

```diff
@@ -0,0 +1,290 @@
+     PodResourceRecommender: logic.CreatePodResourceRecommender(),
+     CheckpointsGCInterval:  checkpointsGCInterval,
+     UseCheckpoints:         useCheckpoints,
+  }.Make()
+}
+
+
+// NewClusterState returns a new ClusterState with no pods.
+func NewClusterState(recommender_name string) *ClusterState {
+  return &ClusterState{
+     RecommenderName:   recommender_name,
+     Pods:              make(map[PodID]*PodState),
+     Vpas:              make(map[VpaID]*Vpa),
+     EmptyVPAs:         make(map[VpaID]time.Time),
+     aggregateStateMap: make(aggregateContainerStatesMap),
+     labelSetMap:       make(labelSetMap),
+  }
+}
+```
+
+Therefore, when loading VPA objects to the `clusterStateFeeder`, it can use the field selector to select VPA CRDs that 
+have `recommenderName` equal to the current clusterState’s `RecommenderName`.
+```golang
+// Fetch VPA objects and load them into the cluster state.
+func (feeder *clusterStateFeeder) LoadVPAs() {
+  // List VPA API objects.
+  allVpaCRDs, err := feeder.vpaLister.List(labels.Everything())
+  if err != nil {
+     klog.Errorf("Cannot list VPAs. Reason: %+v", err)
+     return
+  }
+
+  var vpaCRDs []*vpa_types.VerticalPodAutoscaler
+  for _, vpaCRD := range allVpaCRDs {
+     currentRecommenderName := feeder.clusterState.RecommenderName
+     if (vpaCRD.Spec.RecommenderName != currentRecommenderName) && (vpaCRD.Spec.RecommenderName != "") {
+        klog.V(6).Infof("Ignoring the vpaCRD as its name %v is not equal to the current recommender's name %v", vpaCRD.Spec.RecommenderName, currentRecommenderName)
+        continue
+     }
+     vpaCRDs = append(vpaCRDs, vpaCRD)
+
+  klog.V(3).Infof("Fetched %d VPAs.", len(vpaCRDs))
+  // Add or update existing VPAs in the model.
+  vpaKeys := make(map[model.VpaID]bool)
+
+  …
+
+  feeder.clusterState.ObservedVpas = vpaCRDs
+}
+```
+
+Accordingly, the VPA object yaml should include the `recommenderName` as the default `RecommenderName`.
+```yaml
+apiVersion: "autoscaling.k8s.io/v1"
+kind: VerticalPodAutoscaler
+metadata:
+ name: hamster-vpa
+Spec:
+ recommenderName: default
+ targetRef:
+   apiVersion: "apps/v1"
+   ... ... 
+```
+
+### Deployment Details
+The customized recommender is supposed to be deployed as a separate deployment that is chosen 
+by different sets of VPA objects.Each VPA object is supposed to choose only one recommender at a time.
+The way how the default recommender and the customized recommender are running and interacting with VPA objects 
+are shown in the following drawing.
+
+<img src="images/deployment.png" alt="deployment" width="720" height="360"/>
+
+Though we do not support a VPA object to use multiple recommenders in this proposal, we leave the possibility of necessary  
+changes of using multiple recommenders in the future. Namely, in the future, we may come to multiple recommenders for one VPA
+object and the field of `recommenderName` can be an array instead of a string.
```

Comment:
This might not be so easy - please see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility

So maybe we could instead already make it an array and explicitly state that only the first entry is used for now (as multiple recommenders are not supported with this KEP)?

Reply:
I propose to do as Karol suggests, and have this be an array, with one element only. 
It is indeed not backwards-compatible to change API field type so we would not be able to do this. 

It should be relatively easy to have the admission controller validate that the array has <=1 elements for now.